components/dataHub/adminService.c

```diff
@@ -2223,3 +2223,305 @@ void admin_EndUpdate
+    }
+
+    LE_DEBUG("App name is '%s'.", appName);
+
+    // Now get the app's namespace under the /app namespace.
+    nsRef = resTree_GetEntry(nsRef, appName);
+
+    // Create the Output resource. Need to convert absolute path to relative one.
+    resRef = resTree_GetOutput(nsRef, path + strlen("/app/") + strlen(appName) + strlen("/"), dataType, units);
+    if (resRef == NULL)
+    {
+        LE_ERROR("Failed to create Output '/app/%s/%s'.", resTree_GetEntryName(nsRef), path);
+        return LE_FAULT;
+    }
+
+    return LE_OK;
+}
+
+//--------------------------------------------------------------------------------------------------
+/**
+ * Delete a resource.
+ *
+ * Does nothing if the resource doesn't exist.
+ */
+//--------------------------------------------------------------------------------------------------
+void admin_DeleteResource
+(
+    const char* path
+        ///< [IN] Absolute resource tree path.
+)
+{
+    LE_DEBUG("'%s'", path);
+
+    // If the resource exists, delete it.
+    resTree_EntryRef_t resRef = resTree_FindEntryAtAbsolutePath(path);
+    if (resRef != NULL)
+    {
+        resTree_DeleteIO(resRef);
+    }
+}
+
+//--------------------------------------------------------------------------------------------------
+/**
+ * Set the example value for a JSON-type Input resource.
+ *
+ * Does nothing if the resource is not found, is not an input, or doesn't have a JSON type.
+ */
+//--------------------------------------------------------------------------------------------------
+void admin_SetJsonExample
+(
+    const char* path,
+        ///< [IN] Absolute resource tree path.
+    const char* example
+        ///< [IN] The example JSON value string.
+)
+{
+    resTree_EntryRef_t resRef = resTree_FindEntryAtAbsolutePath(path);
+    if (resRef == NULL)
+    {
+        LE_ERROR("Resource '%s' does not exist.", path);
+    }
+    else if (resTree_GetEntryType(resRef) != ADMIN_ENTRY_TYPE_INPUT)
+    {
+        LE_ERROR("Resource '%s' is not an input.", path);
+    }
+    else if (resTree_GetDataType(resRef) != IO_DATA_TYPE_JSON)
+    {
+        LE_ERROR("Resource '%s' does not have JSON data type.", path);
+    }
+    else
+    {
+        dataSample_Ref_t sampleRef = dataSample_CreateJson(0, example);
+
+        if (sampleRef != NULL)
+        {
+            resTree_SetJsonExample(resRef, sampleRef);
+        }
+    }
+}
+
+//--------------------------------------------------------------------------------------------------
+/**
+ * Mark an Output resource "optional".  (By default, they are marked "mandatory".)
+ */
+//--------------------------------------------------------------------------------------------------
+void admin_MarkOptional
+(
+    const char* path
+        ///< [IN] Absolute resource tree path.
+)
+{
+    resTree_EntryRef_t resRef = resTree_FindEntryAtAbsolutePath(path);
+    if (resRef == NULL)
+    {
+        LE_KILL_CLIENT("Attempt to mark non-existent resource optional at '%s'.", path);
```

Comment:
Should we be using LE_KILL_CLIENT here?  We removed it in some other places to help with stability.

Reply:
Actually this is copy paste from `io_MarkOptional`
I only updated 
`resTree_EntryRef_t resRef = FindResource(path);`
to
`resTree_EntryRef_t resRef = resTree_FindEntryAtAbsolutePath(path);`
I kept the rest of the code as is.

I agree, I will change it to LE_ERROR.